stonesoup/resampler/particle.py

```diff
@@ -83,10 +93,219 @@ def resample(self, particles):
+
+        Returns
+        -------
+        particle state: :class:`~.ParticleState`
+            The particle state after resampling
+        """
+
+        if not isinstance(particles, ParticleState):
+            particles = ParticleState(None, particle_list=particles)
+        if nparts is None:
+            nparts = len(particles)
+
+        log_weights = particles.log_weight
+        weight_order = np.argsort(log_weights, kind='stable')
+        max_log_value = log_weights[weight_order[-1]]
+        with np.errstate(divide='ignore'):
+            cdf = np.log(np.cumsum(np.exp(log_weights[weight_order] - max_log_value)))
+        cdf += max_log_value
+
+        # Pick random points for each of the particles
+        u_j = np.random.rand(nparts)
+
+        # Pick particles that represent the chosen point from the cdf
+        index = weight_order[np.searchsorted(cdf, np.log(u_j))]
+
+        new_particles = particles[index]
+        new_particles.log_weight = np.full((nparts, ), np.log(1/nparts))
+        return new_particles
+
+
+class StratifiedResampler(Resampler):
+    """
+    Traditional style resampler for particle filter. Splits the cdf into N evenly sized
+    subpopulations ('strata'), then independently picks one value from each stratum. Complexity of
+    order O(N).
+
+    """
+
+    def resample(self, particles, nparts=None):
+        """Resample the particles
+
+        Parameters
+        ----------
+        particles : :class:`~.ParticleState` or list of :class:`~.Particle`
+            The particles or particle state to be resampled according to their weights
+        nparts : int
+            The number of particles to be returned from resampling
+
+        Returns
+        -------
+        particle state: :class:`~.ParticleState`
+            The particle state after resampling
+        """
+
+        if not isinstance(particles, ParticleState):
+            particles = ParticleState(None, particle_list=particles)
+        if nparts is None:
+            nparts = len(particles)
+
+        log_weights = particles.log_weight
+        weight_order = np.argsort(log_weights, kind='stable')
+        max_log_value = log_weights[weight_order[-1]]
+        with np.errstate(divide='ignore'):
+            cdf = np.log(np.cumsum(np.exp(log_weights[weight_order] - max_log_value)))
+        cdf += max_log_value
+
+        # Strata lower bounds:
+        s_l = np.arange(nparts) * (1 / nparts)
+
+        # Independently pick a point in each stratum
+        u_j = np.random.uniform(s_l, s_l + (1 / nparts))
+
+        # Pick particles that represent the chosen point from the cdf
+        index = weight_order[np.searchsorted(cdf, np.log(u_j))]
+
+        new_particles = particles[index]
+        new_particles.log_weight = np.full((nparts, ), np.log(1/nparts))
+        return new_particles
+
+
+class ResidualResampler(Resampler):
+    """
+    Traditional resampler. Any particle, p with weight W >= 1/N, will be resampled floor(W_p)
```

Comment:
Is this still a traditional resampler given you made a point of referring to this as a preprocessing method?

Reply:
Changes made to adjust this.